File: cpp/src/global_planning/sampling/prm_star.cpp
#include "navigation/global_planning/sampling/prm_star.hpp"

#include <chrono>

#include "navigation/global_planning/sampling/roadmap_common.hpp"
#include "navigation/global_planning/sampling/sampling_common.hpp"

namespace navigation::global_planning {

core::PlanResult<Point> PrmStarPlanner::plan(SamplingSpace<Point>& space, const Point& start,
                                             const Point& goal, TraceRecorder* recorder) {
  const int num_samples = params_.get_int("num_samples");
  const double gamma = params_.get_float("gamma");

  auto t0 = std::chrono::steady_clock::now();
  Roadmap roadmap;
  int start_idx = roadmap.add_node(start);
  int goal_idx = roadmap.add_node(goal);
  sample_free(space, roadmap, num_samples, recorder);
  // PRM* radius: computed once from the final node count (Karaman & Frazzoli 2011).
  double radius = rgg_radius(gamma, static_cast<int>(roadmap.nodes.size()));
  for (int idx = 1; idx < static_cast<int>(roadmap.nodes.size()); ++idx) {
    connect(space, roadmap, idx, radius, recorder);
  }

  double cost = 0.0;
  int expanded = 0;
  std::vector<Point> path = dijkstra(roadmap, start_idx, goal_idx, recorder, cost, expanded);
  double rt = std::chrono::duration<double>(std::chrono::steady_clock::now() - t0).count();
  bool success = !path.empty();
  if (success && recorder) recorder->path_found(path);

  int nnodes = static_cast<int>(roadmap.nodes.size());
  emit_finished_batch(recorder, success, cost, expanded, nnodes, nnodes, rt);
  core::PlanResult<Point> result;
  result.success = success;
  result.path = std::move(path);
  result.cost = success ? cost : 0.0;
  result.stats = {expanded, nnodes, 0, nnodes};
  return result;
}

}  // namespace navigation::global_planning

File: python/navigation/global_planning/sampling/prm_star.py
"""PRM* — asymptotically optimal PRM (Karaman & Frazzoli 2011).

Identical to PRM except the connection radius is not fixed: it shrinks with the
sample count as r_n = γ·(log n / n)^(1/d) (d = 2). This keeps the expected
neighbour count at Θ(log n), which is exactly what buys almost-sure convergence
to the optimal path (a fixed radius either over-connects or, if too small,
breaks optimality).
"""

from __future__ import annotations

import time

from navigation.core.capabilities import SamplingSpace
from navigation.core.trace import TraceRecorder
from navigation.core.types import PlanResult, PlanStats, Point

from ._roadmap import Roadmap, _RoadmapPlanner, connect, dijkstra
from ._sampling import rgg_radius


class PRMStar(_RoadmapPlanner):
    @property
    def name(self) -> str:
        return "prm_star"

    def plan(
        self,
        space: SamplingSpace[Point],
        start: Point,
        goal: Point,
        recorder: TraceRecorder | None = None,
    ) -> PlanResult[Point]:
        t0 = time.monotonic()
        num_samples = self.params.get_int("num_samples")
        gamma = self.params.get_float("gamma")

        roadmap = Roadmap()
        start_idx = roadmap.add_node(start)
        goal_idx = roadmap.add_node(goal)
        self._sample_free(space, roadmap, num_samples, recorder)
        # PRM* radius: computed once from the final node count (Karaman & Frazzoli 2011).
        radius = rgg_radius(gamma, len(roadmap))
        for idx in range(1, len(roadmap)):
            connect(space, roadmap, idx, radius, recorder)

        path, cost, expanded = dijkstra(roadmap, start_idx, goal_idx, recorder)
        runtime = time.monotonic() - t0
        success = len(path) > 0
        if success and recorder is not None:
            recorder.path_found(path)
        self._finish(recorder, success, cost, expanded, len(roadmap), runtime)
        stats = PlanStats(
            expanded_nodes=expanded, samples=len(roadmap), tree_size=len(roadmap), iterations=0
        )
        return PlanResult(success, path, cost, stats)
